types/redux-api-middleware/index.d.ts

```diff
@@ -0,0 +1,104 @@
+// Type definitions for redux-api-middleware 3.0
+// Project: https://github.com/agraboso/redux-api-middleware
+// Definitions by: Andrew Luca <https://github.com/me>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// TypeScript Version: 2.8
+
+import {
+    Middleware,
+    MiddlewareAPI,
+    Dispatch,
+    AnyAction,
+    applyMiddleware
+} from 'redux';
+
+/**
+ * This module is also a UMD module that exposes a global variable 'ReduxApiMiddleware'
+ * when loaded outside a module loader environment.
+ */
+export as namespace ReduxApiMiddleware;
+
+export type RSAA = '@@redux-api-middleware/RSAA';
+export const RSAA = '@@redux-api-middleware/RSAA';
+
+export function isRSAA(action: object): boolean;
+export function validateRSAA(action: object): string[];
+export function isValidRSAA(action: object): boolean;
+
+export class InvalidRSAA extends Error {
+    name: 'InvalidRSAA';
+    message: 'Invalid RSAA';
+    validationErrors: string[];
+    constructor(validationErrors: string[]);
+}
+
+export class InternalError extends Error {
+    name: 'InternalError';
+    message: string;
+    constructor(message: string);
+}
+
+export class RequestError extends Error {
+    name: 'RequestError';
+    message: string;
+    constructor(message: string);
+}
+
+export class ApiError<T extends object = any> extends Error {
+    name: 'ApiError';
+    status: number;
+    statusText: string;
+    response: T;
+    message: string;
+    constructor(status: number, statusText: string, response: T);
+}
+
+export function getJSON(res: Response): Promise<any> | Promise<void>;
+
+export const apiMiddleware: Middleware;
+
+export interface CreateMiddlewareOptions {
+    ok?: (res: Response) => boolean;
+    fetch?: typeof fetch;
+}
+
+export function createMiddleware(options?: CreateMiddlewareOptions): Middleware;
+
+export interface RSAARequestTypeDescriptor<S = any, P = any, M = any> {
+    type: string | symbol;
+    payload?: (action: RSAAAction, state: S) => P | P;
+    meta?: (action: RSAAAction, state: S) => M | M;
+}
+
+export interface RSAASuccessTypeDescriptor<S = any, P = any, M = any> {
+    type: string | symbol;
+    payload?: (action: RSAAAction, state: S, res: any) => P | P;
+    meta?: (action: RSAAAction, state: S, res: any) => M | M;
+}
+
+export interface RSAAFailureTypeDescriptor<S = any, P = any, M = any> {
+    type: string | symbol;
+    payload?: (action: RSAAAction, state: S) => P | P;
+    meta?: (action: RSAAAction, state: S, res: any) => M | M;
+}
+
+export interface RSAACall<S = any, P = any, M = any> {
+    endpoint: (state: S) => string | string;
```

Comment:
Day 1 of Typescript for me so please correct me if I'm wrong...should this line instead be `endpoint: ((state: S) => string) | string;`

Seems that this is currently saying the function must return a string or a string, instead of, a string or a function that returns a string?

Reply:
Yeap, you're right. Now it seems is evaluated how you said
```
(state: S) => (string | string);
```
and `string | string` equals `string`

Reply:
And not only this one, but all lines that have a type or a function with same exact return type. Also strange behaviour
```ts
// type-test.ts
type Test1 = string | arg => string;
type Test2 = string | (arg: any) => string;
type Test3 = arg => string | string;
type Test4 = (arg: any) => string | string;
```
```bash
andreiluca@trolic:~/Downloads|⇒  npx typescript --noEmit type-test.ts
npx: installed 1 in 2.855s

type-test.ts:1:27 - error TS1005: ';' expected.
1 type Test1 = string | arg => string;
                            ~~

type-test.ts:2:27 - error TS1005: ')' expected.
2 type Test2 = string | (arg: any) => string;
                            ~

type-test.ts:2:32 - error TS1005: ';' expected.
2 type Test2 = string | (arg: any) => string;
                                 ~

type-test.ts:2:34 - error TS1128: Declaration or statement expected.
2 type Test2 = string | (arg: any) => string;
                                   ~~

type-test.ts:3:18 - error TS1005: ';' expected.
3 type Test3 = arg => string | string;
                   ~~
```

Reply:
One solution is to create some aliases
```ts
type TypeResolver<Type, State> = (state: State) => Type;
type TypeOrResolver<Type, State> = Type | TypeResolver<Type, State>;

export interface RSAACall<S = any, P = any, M = any> {
    endpoint: TypeOrResolver<string, S>;
    body?: TypeOrResolver<BodyInit | null, S>;
    headers?: TypeOrResolver<HeadersInit | HeadersInit, S>;
    options?: TypeOrResolver<RequestInit | RequestInit, S>;
}

const test: RSAACall<number> = {
    endpoint: (a: number) => ''
    // ...
}
```

@Mrman Would you like to crete a PR for fixing this? Or I'll fix it.

Reply:
I'll fix in the next day or two. I probably won't go down the alias route, feels like more indirection for something that is fairly simple.